src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -956,27 +958,33 @@ private void BlockDownloaded(uint256 blockHash, Block block)
                 {
                     try
                     {
-                        chainedHeader = this.chainedHeaderTree.FindHeaderAndVerifyBlockIntegrity(block);
+                        chainedHeader = this.chainedHeaderTree.FindHeader(block);
                     }
                     catch (BlockDownloadedForMissingChainedHeaderException)
                     {
                         this.logger.LogTrace("(-)[CHAINED_HEADER_NOT_FOUND]");
                         return;
                     }
-                    catch (IntegrityValidationFailedException integrityException)
-                    {
-                        this.peerBanning.BanAndDisconnectPeer(integrityException.PeerEndPoint, integrityException.BanDurationSeconds, $"Integrity validation failed: {integrityException.Error.Message}");
-
-                        this.logger.LogTrace("(-)[INTEGRITY_VERIFICATION_FAILED]");
-                        return;
-                    }
                 }
                 else
                 {
                     this.logger.LogDebug("Block '{0}' failed to be delivered.", blockHash);
                 }
             }
 
+            if (block != null)
+            {
+                ValidationContext result = this.integrityValidator.VerifyBlockIntegrity(chainedHeader, block);
+
+                if (result.Error != null)
+                {
+                    this.peerBanning.BanAndDisconnectPeer(result.Peer, result.BanDurationSeconds,
```

Comment:
I don't understand where the peer is set in the ValidationContext object.
Not in the tests, but in the normal code.

Reply:
Good find.
The peer is not set anywhere. 

@dangershony

Reply:
However it's not an issue of this PR, it's more about consensus rules engine so let's review this PR assuming it's there.

Reply:
Created an issue for that https://github.com/stratisproject/StratisBitcoinFullNode/issues/2008